### PR TITLE
Online DDL: ddl_strategy session variable and vtctl command line argument

### PR DESCRIPTION
**Pull request by shlomi-noach**
Context: #6782 (comment)
Additional contexts: #6926, #6689
Extends #7042 

In this PR we utilize the new `ddl_strategy` session variable. Consider the following:

```sql

mysql> show create table t \G

Create Table: CREATE TABLE `t` (
  `id` smallint(6) NOT NULL,
  `i` int(11) DEFAULT NULL,
  PRIMARY KEY (`id`)
) ENGINE=InnoDB DEFAULT CHARSET=utf8

mysql> select @@ddl_strategy;
+----------------+
| @@ddl_strategy |
+----------------+
|                |
+----------------+

mysql> alter table t modify id bigint;
Query OK, 0 rows affected (0.03 sec)

mysql> show create table t \G

Create Table: CREATE TABLE `t` (
  `id` bigint(20) NOT NULL,
  `i` int(11) DEFAULT NULL,
  PRIMARY KEY (`id`)
) ENGINE=InnoDB DEFAULT CHARSET=utf8

mysql> set @@ddl_strategy='gh-ost';
Query OK, 0 rows affected (0.00 sec)

mysql> alter table t modify id mediumint;
+--------------------------------------+
| uuid                                 |
+--------------------------------------+
| b87f9824_2814_11eb_bd3d_f875a4d24e90 |
+--------------------------------------+

mysql> show create table t \G

Create Table: CREATE TABLE `t` (
  `id` bigint(20) NOT NULL,
  `i` int(11) DEFAULT NULL,
  PRIMARY KEY (`id`)
) ENGINE=InnoDB DEFAULT CHARSET=utf8

mysql> -- time passes

mysql> show create table t \G

Create Table: CREATE TABLE `t` (
  `id` mediumint(9) NOT NULL,
  `i` int(11) DEFAULT NULL,
  PRIMARY KEY (`id`)
) ENGINE=InnoDB DEFAULT CHARSET=utf8
```

Thus, the choice whether to user "normal" DDL or Online DDL, can depend on the valud of `@@ddl_strategy` session variable, which is either:

- empty (normal DD)
- `gh-ost`
- `pt-osc`

I'm not sure yet how to represent migration options. Would that be `set @@ddl_strategy='gh-ost; max-load=Threads_running=100'`? Seems quite the syntax.

### CODE REVIEW

**Comment by harshit-gangal** [comment]
you can take the input as map in a separate session variable like  `ddl_strategy_settings = 'a=1,b=2'`

File: go/vt/vtgate/engine/ddl.go [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+		}
+	}
+
+	//TODO (shlomi): check session variable
+	return v.NormalDDL.Execute(vcursor, bindVars, wantfields)
+}
+
+//StreamExecute implements the Primitive interface
+func (v *DDL) StreamExecute(vcursor VCursor, bindVars map[string]*query.BindVariable, wantields bool, callback func(*sqltypes.Result) error) error {
+	return vterrors.Errorf(vtrpcpb.Code_INTERNAL, "not reachable") // TODO: shlomi - have no idea if this should work
```
Comment (harshit-gangal) [inline go/vt/vtgate/engine/ddl.go:109]: you can call Execute and pass the result in callback function. The plan is to merge Execute and StreamExecute with streaming only so we need to support then.

File: go/vt/vtgate/engine/ddl.go [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+}
+
+//GetFields implements the Primitive interface
+func (v *DDL) GetFields(vcursor VCursor, bindVars map[string]*query.BindVariable) (*sqltypes.Result, error) {
+	return nil, vterrors.Errorf(vtrpcpb.Code_INTERNAL, "not reachable") // TODO: shlomi - have no idea if this should work
+}
```
Comment (harshit-gangal) [inline go/vt/vtgate/engine/ddl.go:115]: this will not be called. todo can be removed.

**Comment by shlomi-noach** [comment]
> you can take the input as map in a separate session variable like ddl_strategy_settings = 'a=1,b=2'


Thank you. I'm thinnking two variables are difficult to maintain. The user will set one, and forget that the other one is already set from a previous cycle. 

I'm thinking to simplify as `set @@ddl_strategy='gh-ost --max-load=Threads_running=100'` which actually reads like a normal command line execution.

**Comment by shlomi-noach** [comment]
I am _still_ unsure whether this approach of using session variables is any better for our users. Users want to be able to run exact same command on MySQL and on Vitess, such that the Vitess execution will run an online DDL.

So, the `ALTER TABLE` is fine. But how will they be able to run `set @@ddl_strategy='gh-ost';` on their MySQL server? I feel like we;re back to square one.

**Review by harshit-gangal — APPROVED** [review]

File: go/vt/vtgate/engine/ddl.go (hunk 1 above)
Comment (shlomi-noach) [reply go/vt/vtgate/engine/ddl.go]: Implemented

**Comment by harshit-gangal** [comment]
> I am _still_ unsure whether this approach of using session variables is any better for our users. Users want to be able to run exact same command on MySQL and on Vitess, such that the Vitess execution will run an online DDL.
> 
> So, the `ALTER TABLE` is fine. But how will they be able to run `set @@ddl_strategy='gh-ost';` on their MySQL server? I feel like we;re back to square one.

Currently there are 2 ways to execute DDL in Vitess.

1. ApplySchema command in vtctl: we can add online-ddl option there. That can internally handle how to call things
2. Through VTGate
* via some MySQL clients (cli or UI) - we can drive the experience there to execute the set statements first to enable online_ddl and then execute the ddl query.
* via programatically - this we can give flag on vtgate that can be set at startup time so that all ddl are treated with that strategy and user will have option to change it with set statement.

**Comment by harshit-gangal** [comment]
There are few places that would need additional test
1. engine/ddl 
2. executor
3. end to end test if that are not present.

**Comment by shlomi-noach** [comment]
> ApplySchema command in vtctl: we can add online-ddl option there. That can internally handle how to call things

How would it look like in `ApplySchema` in your opinion?

> we can drive the experience there

How do you mean?

> we can give flag on vtgate 

Yes, I was also thinking about that. One problem I have now is that, say we have a command line flag that sets to `gh-ost`. Now the user overrides with `set session ddl_strategy=''`. How do we _know_ that the user set the strategy to <empty> (therefore, we should run a normal alter table) as opposed to not setting the session variable (thus, we should run gh-ost)?

**Comment by shlomi-noach** [comment]
OK, this PR now supports:

- `ddl_strategy` session variable. Use as follows:
```sql
set session ddl_strategy='gh-ost';
set @@ddl_strategy='gh-ost';
set @@ddl_strategy='gh-ost --max-load=Threads_running=100';
set @@ddl_strategy='';
select @@ddl_strategy;
```
- `vtgate` command line argument `--default-ddl-strategy`
  the session variable `@@ddl_strategy` defaults to that value
- `vtctl ApplySchema` now supports `--ddl_strategy` command line argument
  in exact same format as `@@ddl_strategy`

`onlineDDL` endtoend tests cover all variations (vtgate, vtctl, with/out session variables, with/out params)

**A Followup PR** will remove the `WITH 'gh-ost'` syntax from parser, ast etc. It's too much for this single PR.

Documentation to follow.

**Comment by shlomi-noach** [comment]
@harshit-gangal I've made changes since your last review; see https://github.com/vitessio/vitess/pull/7045#issuecomment-731725891

File: go/vt/schema/online_ddl_test.go [hunk 3]
```diff
@@ -27,24 +29,38 @@ func TestCreateUUID(t *testing.T) {
...
-		strategy, err := ValidateDDLStrategy("gh-ost")
-		assert.NoError(t, err)
-		assert.Equal(t, DDLStrategyGhost, strategy)
+func TestParseDDLStrategy(t *testing.T) {
+	tt := []struct {
+		strategyVariable string
+		strategy         sqlparser.DDLStrategy
+		options          string
+		err              error
+	}{
```
Comment (harshit-gangal) [inline go/vt/schema/online_ddl_test.go:38]: add an invalid parsing test.

File: go/vt/vtgate/engine/ddl.go [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+	vtrpcpb "vitess.io/vitess/go/vt/proto/vtrpc"
+	"vitess.io/vitess/go/vt/schema"
+	"vitess.io/vitess/go/vt/sqlparser"
+	"vitess.io/vitess/go/vt/vterrors"
+	"vitess.io/vitess/go/vt/vtgate/vindexes"
+)
+
+var _ Primitive = (*DDL)(nil)
+
+//DDL represents the a DDL statement, either normal or online DDL
```
Comment (harshit-gangal) [inline go/vt/vtgate/engine/ddl.go:31]: nit: convention in the code base is to have space between // and comment.

File: go/vt/vtgate/vtgate.go [hunk 5]
```diff
@@ -60,6 +61,7 @@ var (
 	_                  = flag.Bool("disable_local_gateway", false, "deprecated: if specified, this process will not route any queries to local tablets in the local cell")
 	maxMemoryRows      = flag.Int("max_memory_rows", 300000, "Maximum number of rows that will be held in memory for intermediate results as well as the final result.")
 	warnMemoryRows     = flag.Int("warn_memory_rows", 30000, "Warning threshold for in-memory results. A row count higher than this amount will cause the VtGateWarnings.ResultsExceeded counter to be incremented.")
+	defaultDDLStrategy = flag.String("default-ddl-strategy", "", "Set default strategy for DDL statements. Override with @@ddl_strategy session variable")
```
Comment (harshit-gangal) [inline go/vt/vtgate/vtgate.go:64]: I think we can drop the word `default` from here. This is how some of the other startup parameters behave.

**Review by harshit-gangal — APPROVED** [review]
Few nitpick comments. Everything else looks good to me.

File: go/vt/schema/online_ddl_test.go (hunk 3 above)
Comment (shlomi-noach) [reply go/vt/schema/online_ddl_test.go:38]: There is one, at the end of the function

File: go/vt/vtgate/engine/ddl.go (hunk 4 above)
Comment (shlomi-noach) [reply go/vt/vtgate/engine/ddl.go:31]: fixed

File: go/vt/vtgate/vtgate.go (hunk 5 above)
Comment (shlomi-noach) [reply go/vt/vtgate/vtgate.go:64]: fixed

**Review by harshit-gangal — APPROVED** [review]

**Comment by shlomi-noach** [comment]
Documentation: https://github.com/vitessio/website/pull/598

**Pull request merged**